File: Labs/Lab-1/BST.java
import java.util.LinkedList;
import java.util.Queue;
import java.util.Scanner;
class Node {
    int data;
    Node left, right;
    Node(int item) {
        data = item;
        left = right = null;
    }
}
class BST {
    Node root;
    BST() {
        root = null;
    }
    Node insert(Node root, int data) {
        if (root == null) {
            root = new Node(data);
            return root;
        }
        if (data < root.data) {
            root.left = insert(root.left, data);
        }
        else if (data > root.data) {
            root.right = insert(root.right, data);
        }
        return root;
    }
    void inorder(Node root) {
        if (root != null) {
            inorder(root.left);
            System.out.print(root.data + " ");
            inorder(root.right);
        }
    }
    
    Node deleteNode(Node root, int key) {
        if (root == null) {
            return root;
        }
        if (key < root.data) {
            root.left = deleteNode(root.left, key);
        }
        else if (key > root.data) {
            root.right = deleteNode(root.right, key);
        }
        else {
            if (root.left == null)
                return root.right;
            else if (root.right == null)
                return root.left;
            root.data = minValue(root.right);
            root.right = deleteNode(root.right, root.data);
        }
        return root;
    }
    int minValue(Node root) {
        int minv = root.data;
        while (root.left != null) {
            minv = root.left.data;
            root = root.left;
        }
        return minv;
    }
    Node searchNode(Node root, int key) {
        if (root == null || root.data == key) {
            return root;
        }
        if (root.data > key) {
            return searchNode(root.left, key);
        }
        return searchNode(root.right, key);
    }
    Node BFS(Node root) {
        if (root == null) {
            return root;
        }
        Queue<Node> q = new LinkedList<Node>();
        q.add(root);
        while (!q.isEmpty()) {
            Node temp = q.poll();
            System.out.print(temp.data + " ");
            if (temp.left != null) {
                q.add(temp.left);
            }
            if (temp.right != null) {
                q.add(temp.right);
            }
        }
        return root;
    }

    public static void main(String[] args) {
        System.out.println("Binary Search Tree Operations:");
        BST tree = new BST();
        Scanner sc = new Scanner(System.in);
        boolean exit = false;
        while (!exit) { 
            System.out.println("Menu:");
            System.out.println("1. Insert");
            System.out.println("2. Inorder Traversal");
            System.out.println("3. Delete Node");
            System.out.println("4. Search Node");
            System.out.println("5. BFS");
            System.out.println("6. Exit");
            System.out.print("Enter your choice: ");
            int choice = sc.nextInt(); 
            switch (choice) {
                case 1:
                    System.out.print("Enter value to insert: ");
                    int valToInsert = sc.nextInt();
                    tree.root = tree.insert(tree.root, valToInsert);
                    System.out.println("Inserted " + valToInsert);
                    break;  
                case 2:
                    System.out.print("Inorder Traversal: ");
                    tree.inorder(tree.root);
                    System.out.println();
                    break;
                case 3:
                    System.out.print("Enter value to delete: ");
                    int valToDelete = sc.nextInt();
                    tree.root = tree.deleteNode(tree.root, valToDelete);
                    System.out.println("Deleted " + valToDelete);
                    break;
                case 4:
                    System.out.print("Enter value to search: ");
                    int valToSearch = sc.nextInt();
                    Node searchResult = tree.searchNode(tree.root, valToSearch);
                    if (searchResult != null) {
                        System.out.println("Found " + valToSearch + " at node with address: " + searchResult);
                    } else {
                        System.out.println("Not found " + valToSearch);
                    }
                    break;
                case 5:
                    System.out.print("BFS Traversal: ");
                    tree.BFS(tree.root);
                    System.out.println();
                    break;
                case 6:
                    System.out.println("Exiting...");
                    exit = true;
                    break;
                default:
                    System.out.println("Invalid choice! Please try again.");
            }
        }
        sc.close();
    }
}

File: Practise/BST.cpp
#include <iostream>
#include <queue>
using namespace std;
struct Node{
    int data;
    Node* left;
    Node* right;
};
void insert(Node *&root, int key){
    if(root==nullptr){
        root=new Node();
        root->data=key;
        root->left=nullptr;
        root->right=nullptr;
        return;
    }
    if(key<root->data){
        insert(root->left,key);
    }else{
        insert(root->right,key);
    }   
}
Node *findMin(Node *root){
    while(root&&root->left!=nullptr){
        root=root->left;
    }
    return root;
}
void deleteNode(Node *&root, int key)
{
    if(root==nullptr){
        cout<<"Element"<<key<<"not found!"<<endl;
        return;
    }
    if(key<root->data){
        deleteNode(root->left,key);
    }else if(key>root->data){
        deleteNode(root->right,key);
    }else{
        if(root->left==nullptr){
            Node *temp=root->right;
            delete root;
            root=temp;
        }
        else if(root->right==nullptr){
            Node *temp=root->left;
            delete root;
            root=temp;
        }
        else{
            Node *temp=findMin(root->right);
            root->data=temp->data;
            deleteNode(root->right,temp->data);
        }
        cout<<"Deleted "<<key<<endl;
    }
}
void search(Node *root,int key){
    if(root==nullptr){
        cout<<"Key Not Found!"<<endl;
        return;
    }
    if(root->data == key){
        cout<<"Key "<<key<<" found in the tree!"<<endl;
    }else if(key<root->data){
        search(root->left,key);
    }else{
        search(root->right,key);
    }
}
void displayInOrder(Node *root){
    if(root==nullptr){
        return;
    }
    displayInOrder(root->left);
    cout<<root->data<<" ";
    displayInOrder(root->right);
}
void printCurrentLevel(Node* root,int level) {
    if(root == nullptr){
        return;
    }
    if(level==1){
        cout<<root->data<<" ";
    }else if(level > 1){
        printCurrentLevel(root->left,level-1);
        printCurrentLevel(root->right,level-1);
    }
}
int getHeight(Node *root){
    if(root==nullptr){
        return -1;
    }
    return 1 + max(getHeight(root->left), getHeight(root->right));
}
void BFS(Node* root){
    int h=getHeight(root);
    for(int i=1;i<=h+1;i++){
        printCurrentLevel(root,i);
    }
}
int main(){
    Node *root = nullptr;
    int ch, in;
    while (true)
    {
        cout << "\nChoice Menu";
        cout << "\n1. Insert\n2. Delete\n3. Search\n4. Display\n5. BFS\n6. Height Check\n7. Exit";
        cout << "\nEnter choice: ";
        cin >> ch;
        switch (ch)
        {
        case 1:
            cout << "Enter value to insert: ";
            cin >> in;
            insert(root, in);
            break;
        case 2:
            cout << "Enter value to delete: ";
            cin >> in;
            deleteNode(root, in);
            break;
        case 3:
            cout << "Enter value to search: ";
            cin >> in;
            search(root, in);
            break;
        case 4:
            cout << "In-order: ";
            displayInOrder(root);
            cout << endl;
            break;
        case 5:
            cout << "BFS (Level wise): ";
            BFS(root);
            cout << endl;
            break;
        case 6:
            cout << "Current Height: " << getHeight(root) << endl;
            break;
        case 7:
            cout << "Program ended" << endl;
            return 0;
        default:
            cout << "Invalid choice" << endl;
        }
    }
    return 0;
}